components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/handler/request/PostAuthnMissingChallengeQuestionsHandler.java

```diff
@@ -0,0 +1,420 @@
+     * @return boolean value indicating whether the user has already provided challenge questions.
+     */
+    private boolean isChallengeQuestionsProvided(AuthenticatedUser user) {
+
+        try {
+            String userName = user.getUserName();
+            int tenantId = Utils.getTenantId(MultitenantUtils.getTenantDomain(userName));
+            UserStoreManager userStoreManager =
+                    IdentityRecoveryServiceDataHolder.getInstance().getRealmService()
+                            .getTenantUserRealm(tenantId)
+                            .getUserStoreManager();
+            Map<String, String> claimsMap = userStoreManager
+                    .getUserClaimValues(userName, new String[]{IdentityRecoveryConstants.CHALLENGE_QUESTION_URI},
+                            UserCoreConstants.DEFAULT_PROFILE);
+            String claimValue = claimsMap.get(IdentityRecoveryConstants.CHALLENGE_QUESTION_URI);
+            return claimValue != null;
+
+        } catch (IdentityException e) {
+            log.error("Identity exception occurred while retrieving tenant ID for the user :" + user.getUserName(),
+                    e);
+
+        } catch (UserStoreException e) {
+            log.error("User store exception occurred while retrieving user store for the user :" + user.getUserName(),
+                    e);
+        }
+        return false;
+    }
+
+    /**
+     * Returns a list of challenge questions for a given user.
+     *
+     * @param user Authenticated User.
+     * @return List of ChallengeQuestions.
+     */
+    private List<ChallengeQuestion> getChallengeQuestions(AuthenticatedUser user) {
+        String tenantDomain = MultitenantUtils.getTenantDomain(user.getUserName());
+
+        try {
+            return ChallengeQuestionManager.getInstance().getAllChallengeQuestions(tenantDomain);
+        } catch (IdentityRecoveryServerException e) {
+            log.error("Identity recovery server error occurred for user:" + user.getUserName(), e);
+            return null;
+        }
+    }
+
+    /**
+     * Set the challenge questions for the user by calling ChallengeQuestionManger.
+     *
+     * @param user                 Authenticated User.
+     * @param userChallengeAnswers Array of UserChallengeAnswer
+     */
+    private void setChallengeQuestionAnswers(User user, UserChallengeAnswer[] userChallengeAnswers) {
+
+        try {
+            ChallengeQuestionManager.getInstance().setChallengesOfUser(user, userChallengeAnswers);
+        } catch (IdentityRecoveryException e) {
+            log.error("Unable to save challenge question answers for user : " + user.getUserName(), e);
+        }
+    }
+
+    /**
+     * Returns an array of UserChallengeAnswer from constructed from the servlet request parameters
+     *
+     * @param servletRequest HTTP Servlet Request.
+     * @return challengeQuestionList.
+     */
+    private UserChallengeAnswer[] retrieveChallengeQuestionAnswers(HttpServletRequest servletRequest,
```

Comment:
please rename the list>>challengeQuestionList

Reply:
fixed